File: platform/fastapi_server/config/lifecycle/535_s3_api.lifecycle.py
"""
AWS S3 Storage API Lifecycle Hook for FastAPI

Registers versioned provider routes for S3 CRUD operations
under the /~/api/rest/{api_release_date}/providers/s3_api prefix.
Bucket is provided as a URL path parameter.

Loading Order: 535 (after GitHub SDK audit at 530)

Environment Variables (for S3 client config):
    AWS_S3_REGION        - AWS region (default: us-east-1)
    AWS_S3_ENDPOINT      - Custom endpoint URL (for LocalStack, MinIO)
    AWS_S3_ACCESS_KEY    - AWS access key ID
    AWS_S3_SECRET_KEY    - AWS secret access key

Registered endpoints (prefix: /~/api/rest/{api_release_date}/providers/s3_api):
    GET    /health                                           - Health check
    GET    /v1/buckets                                       - List all buckets
    POST   /v1/buckets/{bucket}                              - Create bucket
    HEAD   /v1/buckets/{bucket}                              - Check bucket exists
    GET    /v1/buckets/{bucket}/objects                      - List objects (query: prefix, cursor, maxKeys)
    POST   /v1/buckets/{bucket}/objects                      - Put object (body: { key, data, contentType? })
    GET    /v1/buckets/{bucket}/objects/by-key               - Get object (query: key)
    PUT    /v1/buckets/{bucket}/objects/by-key               - Update object (query: key, body: data)
    DELETE /v1/buckets/{bucket}/objects/by-key               - Delete object (query: key)
    HEAD   /v1/buckets/{bucket}/objects/by-key               - Check object exists / metadata (query: key)
    GET    /v1/buckets/{bucket}/objects/metadata             - Get object metadata (query: key)
"""

import logging
import time
from typing import Any

from botocore.config import Config as BotoConfig
from fastapi import APIRouter, FastAPI, Query, Request
from fastapi.responses import JSONResponse, Response

from aws_s3_client import config_from_env

logger = logging.getLogger("lifecycle.s3_api")

VENDOR = "s3_api"
VENDOR_VERSION = "v1.0.0"


def _get_yaml_s3(app: FastAPI) -> dict | None:
    """Get storage.s3 section from AppYamlConfig, or None if unavailable."""
    app_config = getattr(getattr(app, "state", None), "config", None)
    if app_config and hasattr(app_config, "get_nested"):
        return app_config.get_nested("storage", "s3")
    return None


def _build_boto_config(config: dict) -> BotoConfig:
    """Build a botocore Config from the S3 config dict."""
    kwargs: dict = {
        "connect_timeout": config.get("connect_timeout", 10),
        "read_timeout": config.get("read_timeout", 60),
        "retries": {"max_attempts": config.get("max_retries", 3)},
    }
    proxy_url = config.get("proxy_url")
    if proxy_url:
        kwargs["proxies"] = {"http": proxy_url, "https": proxy_url}
    return BotoConfig(**kwargs)


async def _create_s3_client(app: FastAPI):
    """Create an aiobotocore S3 client from resolved config."""
    from aiobotocore.session import get_session

    yaml_s3 = _get_yaml_s3(app)
    config = config_from_env(yaml_config=yaml_s3)
    session = get_session()
    client_kwargs: dict = {
        "region_name": config["region"],
        "config": _build_boto_config(config),
        "verify": config.get("verify_ssl", False),
    }
    if config["endpoint_url"]:
        client_kwargs["endpoint_url"] = config["endpoint_url"]
    if config["aws_access_key_id"] and config["aws_secret_access_key"]:
        client_kwargs["aws_access_key_id"] = config["aws_access_key_id"]
        client_kwargs["aws_secret_access_key"] = config["aws_secret_access_key"]

    client = await session.create_client("s3", **client_kwargs).__aenter__()
    return client, config


def _is_not_found(exc: Exception) -> bool:
    """Check if an exception is a 404 / NoSuchKey type error."""
    err_str = str(exc).lower()
    return "nosuchkey" in err_str or "not found" in err_str or "404" in err_str


async def onStartup(app: FastAPI, config: dict[str, Any]) -> None:
    """Startup hook -- Register S3 provider routes."""
    logger.info("Starting s3_api lifecycle hook...")
    try:
        logger.info("Initializing S3 API...")

        s3_api_release_date = app.state.config.get_nested(
            "api_release_date", "contract_snapshot_date", "provider_s3"
        )
        PREFIX = f"/~/api/rest/{s3_api_release_date}/providers/{VENDOR}"
        logger.debug("S3 API prefix: %s", PREFIX)

        # Test connectivity
        configured = False
        s3_config: dict = {}
        try:
            yaml_s3 = _get_yaml_s3(app)
            s3_config = config_from_env(yaml_config=yaml_s3)
            # Quick connectivity test
            client, _ = await _create_s3_client(app)
            await client.__aexit__(None, None, None)
            configured = True
            logger.info(
                "S3 client verified: region=%s endpoint=%s",
                s3_config.get("region"),
                s3_config.get("endpoint_url") or "(default)",
            )
        except Exception as err:
            logger.warning(
                "S3 client creation failed -- v1 routes will NOT be registered. %s", err
            )

        router = APIRouter(prefix=PREFIX)

        @router.get("/health")
        async def s3_health():
            return JSONResponse(content={
                "status": "ok",
                "vendor": VENDOR,
                "vendor_version": VENDOR_VERSION,
                "configured": configured,
                "region": s3_config.get("region"),
                "endpointUrl": s3_config.get("endpoint_url"),
            })

        if configured:
            v1 = APIRouter(prefix="/v1")

            # ── List all buckets ──
            @v1.get("/buckets")
            async def list_buckets(request: Request):
                start = time.monotonic()
                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    response = await client.list_buckets()
                    buckets = [
                        {
                            "name": b.get("Name"),
                            "creationDate": b["CreationDate"].isoformat()
                            if b.get("CreationDate")
                            else None,
                        }
                        for b in response.get("Buckets", [])
                    ]
                    return {
                        "status": "ok",
                        "latency_ms": round((time.monotonic() - start) * 1000),
                        "buckets": buckets,
                    }
                except Exception as exc:
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Create bucket ──
            @v1.post("/buckets/{bucket}")
            async def create_bucket(bucket: str, request: Request):
                start = time.monotonic()
                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    await client.create_bucket(Bucket=bucket)
                    return JSONResponse(
                        status_code=201,
                        content={
                            "status": "ok",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "bucket": bucket,
                            "created": True,
                        },
                    )
                except Exception as exc:
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Check bucket exists ──
            @v1.head("/buckets/{bucket}")
            async def head_bucket(bucket: str, request: Request):
                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    await client.head_bucket(Bucket=bucket)
                    return Response(status_code=200)
                except Exception:
                    return Response(status_code=404)
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── List objects in bucket ──
            @v1.get("/buckets/{bucket}/objects")
            async def list_objects(
                bucket: str,
                request: Request,
                prefix: str = "",
                cursor: str = "",
                maxKeys: int = 50,
            ):
                start = time.monotonic()
                max_keys_val = min(200, max(1, maxKeys))
                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    params: dict = {"Bucket": bucket, "MaxKeys": max_keys_val}
                    if prefix:
                        params["Prefix"] = prefix
                    if cursor:
                        params["ContinuationToken"] = cursor

                    response = await client.list_objects_v2(**params)

                    objects = [
                        {
                            "key": obj.get("Key"),
                            "size": obj.get("Size"),
                            "lastModified": obj["LastModified"].isoformat()
                            if obj.get("LastModified")
                            else None,
                            "storageClass": obj.get("StorageClass"),
                        }
                        for obj in response.get("Contents", [])
                    ]
                    common_prefixes = [
                        p.get("Prefix") for p in response.get("CommonPrefixes", [])
                    ]

                    return {
                        "status": "ok",
                        "latency_ms": round((time.monotonic() - start) * 1000),
                        "bucket": bucket,
                        "prefix": prefix or None,
                        "nextCursor": response.get("NextContinuationToken"),
                        "isTruncated": response.get("IsTruncated", False),
                        "keyCount": response.get("KeyCount", 0),
                        "objects": objects,
                        "commonPrefixes": common_prefixes,
                    }
                except Exception as exc:
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Put object (create/update) ──
            @v1.post("/buckets/{bucket}/objects")
            async def put_object(bucket: str, request: Request):
                start = time.monotonic()
                client = None
                try:
                    body = await request.json()
                    key = body.get("key")
                    data = body.get("data", {})
                    content_type = body.get("contentType", "application/json")

                    if not key:
                        return JSONResponse(
                            status_code=400,
                            content={"status": "error", "error": "key is required in request body"},
                        )

                    encoded = data if isinstance(data, str) else __import__("json").dumps(data)

                    client, _ = await _create_s3_client(request.app)
                    await client.put_object(
                        Bucket=bucket,
                        Key=key,
                        Body=encoded.encode("utf-8"),
                        ContentType=content_type,
                    )

                    return JSONResponse(
                        status_code=201,
                        content={
                            "status": "ok",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "bucket": bucket,
                            "key": key,
                            "size_bytes": len(encoded.encode("utf-8")),
                        },
                    )
                except Exception as exc:
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Get object by key ──
            @v1.get("/buckets/{bucket}/objects/by-key")
            async def get_object(
                bucket: str,
                request: Request,
                key: str = Query(..., description="S3 object key"),
            ):
                start = time.monotonic()
                if not key:
                    return JSONResponse(
                        status_code=400,
                        content={"status": "error", "error": 'query parameter "key" is required'},
                    )

                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    response = await client.get_object(Bucket=bucket, Key=key)
                    content = await response["Body"].read()
                    content_str = content.decode("utf-8")

                    import json as _json
                    try:
                        parsed = _json.loads(content_str)
                    except (ValueError, _json.JSONDecodeError):
                        parsed = content_str

                    return {
                        "status": "ok",
                        "latency_ms": round((time.monotonic() - start) * 1000),
                        "bucket": bucket,
                        "key": key,
                        "contentType": response.get("ContentType"),
                        "contentLength": response.get("ContentLength", 0),
                        "data": parsed,
                    }
                except Exception as exc:
                    if _is_not_found(exc):
                        return JSONResponse(
                            status_code=404,
                            content={"status": "error", "error": f"Object not found: {key}"},
                        )
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Update object by key ──
            @v1.put("/buckets/{bucket}/objects/by-key")
            async def update_object(
                bucket: str,
                request: Request,
                key: str = Query(..., description="S3 object key"),
            ):
                start = time.monotonic()
                if not key:
                    return JSONResponse(
                        status_code=400,
                        content={"status": "error", "error": 'query parameter "key" is required'},
                    )

                client = None
                try:
                    body = await request.body()
                    content_type = request.headers.get("content-type", "application/json")

                    client, _ = await _create_s3_client(request.app)
                    await client.put_object(
                        Bucket=bucket,
                        Key=key,
                        Body=body,
                        ContentType=content_type,
                    )

                    return {
                        "status": "ok",
                        "latency_ms": round((time.monotonic() - start) * 1000),
                        "bucket": bucket,
                        "key": key,
                        "size_bytes": len(body),
                    }
                except Exception as exc:
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Delete object by key ──
            @v1.delete("/buckets/{bucket}/objects/by-key")
            async def delete_object(
                bucket: str,
                request: Request,
                key: str = Query(..., description="S3 object key"),
            ):
                start = time.monotonic()
                if not key:
                    return JSONResponse(
                        status_code=400,
                        content={"status": "error", "error": 'query parameter "key" is required'},
                    )

                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    await client.delete_object(Bucket=bucket, Key=key)

                    return {
                        "status": "ok",
                        "latency_ms": round((time.monotonic() - start) * 1000),
                        "bucket": bucket,
                        "key": key,
                        "deleted": True,
                    }
                except Exception as exc:
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Head object (exists + metadata) ──
            @v1.head("/buckets/{bucket}/objects/by-key")
            async def head_object(
                bucket: str,
                request: Request,
                key: str = Query(..., description="S3 object key"),
            ):
                if not key:
                    return Response(status_code=400)

                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    response = await client.head_object(Bucket=bucket, Key=key)
                    headers = {
                        "x-s3-content-type": response.get("ContentType", ""),
                        "x-s3-content-length": str(response.get("ContentLength", 0)),
                        "x-s3-last-modified": response["LastModified"].isoformat()
                        if response.get("LastModified")
                        else "",
                        "x-s3-etag": response.get("ETag", ""),
                        "x-s3-storage-class": response.get("StorageClass", ""),
                    }
                    return Response(status_code=200, headers=headers)
                except Exception:
                    return Response(status_code=404)
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            # ── Get object metadata (GET version) ──
            @v1.get("/buckets/{bucket}/objects/metadata")
            async def get_object_metadata(
                bucket: str,
                request: Request,
                key: str = Query(..., description="S3 object key"),
            ):
                start = time.monotonic()
                if not key:
                    return JSONResponse(
                        status_code=400,
                        content={"status": "error", "error": 'query parameter "key" is required'},
                    )

                client = None
                try:
                    client, _ = await _create_s3_client(request.app)
                    response = await client.head_object(Bucket=bucket, Key=key)

                    return {
                        "status": "ok",
                        "latency_ms": round((time.monotonic() - start) * 1000),
                        "bucket": bucket,
                        "key": key,
                        "contentType": response.get("ContentType"),
                        "contentLength": response.get("ContentLength", 0),
                        "lastModified": response["LastModified"].isoformat()
                        if response.get("LastModified")
                        else None,
                        "eTag": response.get("ETag"),
                        "storageClass": response.get("StorageClass"),
                        "metadata": response.get("Metadata", {}),
                        "serverSideEncryption": response.get("ServerSideEncryption"),
                    }
                except Exception as exc:
                    if _is_not_found(exc):
                        return JSONResponse(
                            status_code=404,
                            content={"status": "error", "error": f"Object not found: {key}"},
                        )
                    return JSONResponse(
                        status_code=500,
                        content={
                            "status": "error",
                            "latency_ms": round((time.monotonic() - start) * 1000),
                            "error": "S3 operation failed",
                        },
                    )
                finally:
                    if client:
                        await client.__aexit__(None, None, None)

            router.include_router(v1)

        app.include_router(router)

        if configured:
            logger.info("S3 API initialized -- routes registered at %s/*", PREFIX)
        else:
            logger.info(
                "Health endpoint registered at %s/health (v1 routes skipped -- no S3 client)",
                PREFIX,
            )

        logger.info("s3_api lifecycle hook completed successfully")
    except Exception as exc:
        logger.error("s3_api lifecycle hook failed: %s", exc, exc_info=True)
        raise


async def onShutdown(app: FastAPI, config: dict[str, Any]) -> None:
    """Shutdown hook."""
    logger.info("s3_api shutdown complete")

File: platform/fastify_server/config/lifecycle/535.s3_api.lifecycle.mjs
/**
 * AWS S3 Storage API Lifecycle Hook for Fastify
 *
 * Registers versioned provider routes for S3 CRUD operations
 * under the /~/api/rest/{api_release_date}/providers/s3_api prefix.
 * Bucket is provided as a URL path parameter.
 *
 * Loading Order: 535 (after GitHub SDK audit at 530)
 *
 * Environment Variables (for S3 client config):
 *   AWS_S3_REGION        - AWS region (default: us-east-1)
 *   AWS_S3_ENDPOINT      - Custom endpoint URL (for LocalStack, MinIO)
 *   AWS_S3_ACCESS_KEY    - AWS access key ID
 *   AWS_S3_SECRET_KEY    - AWS secret access key
 *
 * Registered endpoints (prefix: /~/api/rest/{api_release_date}/providers/s3_api):
 *   GET    /health                                           - Health check
 *   GET    /v1/buckets                                       - List all buckets
 *   POST   /v1/buckets/:bucket                               - Create bucket
 *   HEAD   /v1/buckets/:bucket                               - Check bucket exists
 *   GET    /v1/buckets/:bucket/objects                       - List objects (query: prefix, cursor, maxKeys)
 *   POST   /v1/buckets/:bucket/objects                       - Put object (body: { key, data, contentType? })
 *   GET    /v1/buckets/:bucket/objects/by-key                - Get object (query: key)
 *   PUT    /v1/buckets/:bucket/objects/by-key                - Update object (query: key, body: data)
 *   DELETE /v1/buckets/:bucket/objects/by-key                - Delete object (query: key)
 *   GET    /v1/buckets/:bucket/objects/metadata              - Get object metadata (query: key)
 */

import {
  S3Client,
  ListBucketsCommand,
  HeadBucketCommand,
  CreateBucketCommand,
  ListObjectsV2Command,
  PutObjectCommand,
  GetObjectCommand,
  DeleteObjectCommand,
  HeadObjectCommand,
} from '@aws-sdk/client-s3';
import { NodeHttpHandler } from '@smithy/node-http-handler';
import { configFromEnv } from 'aws-s3-client';

const VENDOR = 's3_api';
const VENDOR_VERSION = 'v1.0.0';

/**
 * Build an S3Client from resolved config.
 * @param {import('aws-s3-client').SDKConfig} config
 * @returns {Promise<S3Client>}
 */
async function buildS3Client(config) {
  const clientConfig = { region: config.region };

  if (config.endpointUrl) {
    clientConfig.endpoint = config.endpointUrl;
  }
  clientConfig.forcePathStyle = config.forcePathStyle ?? !!config.endpointUrl;

  if (config.awsAccessKeyId && config.awsSecretAccessKey) {
    clientConfig.credentials = {
      accessKeyId: config.awsAccessKeyId,
      secretAccessKey: config.awsSecretAccessKey,
    };
  }

  if (config.proxyUrl) {
    const { HttpsProxyAgent } = await import('https-proxy-agent');
    const agent = new HttpsProxyAgent(config.proxyUrl);
    clientConfig.requestHandler = new NodeHttpHandler({
      httpAgent: agent,
      httpsAgent: agent,
    });
  }

  return new S3Client(clientConfig);
}

/**
 * Startup hook -- Register S3 provider routes.
 * @param {import('fastify').FastifyInstance} server
 * @param {object} config
 */
export async function onStartup(server, config) {
  server.log.info('[lifecycle:s3_api] Initializing S3 API...');

  try {
    const apiReleaseDate = server.config?.getNested?.(
      ['api_release_date', 'contract_snapshot_date', 'provider_s3'],
    ) ?? '02-01-2026';
    const PREFIX = `/~/api/rest/${apiReleaseDate}/providers/${VENDOR}`;
    server.log.info({ PREFIX, apiReleaseDate, VENDOR }, '[lifecycle:s3_api] Resolved API route prefix');

    // Resolve S3 config (credentials, endpoint, region) — bucket comes from URL
    const yamlS3 = server.config?.getNested?.(['storage', 's3']) ?? null;
    const s3Config = configFromEnv(undefined, yamlS3);

    let configured = false;
    let s3Client;
    try {
      s3Client = await buildS3Client(s3Config);
      configured = true;

      if (!server.hasDecorator('s3Client')) server.decorate('s3Client', s3Client);
      server.log.info(
        { region: s3Config.region, endpointUrl: s3Config.endpointUrl || '(default)' },
        '[lifecycle:s3_api] S3 client initialized',
      );
    } catch (err) {
      server.log.warn({ err }, '[lifecycle:s3_api] S3 client creation failed -- v1 routes will NOT be registered.');
    }

    /** Helper: run an operation with the S3 client, catch and format errors. */
    function isNotFoundError(err) {
      const errStr = String(err).toLowerCase();
      return errStr.includes('nosuchkey') || errStr.includes('not found') || errStr.includes('404');
    }

    await server.register(
      async function s3ApiRoutes(scope) {
        // Health -- always registered
        scope.get('/health', async () => ({
          status: 'ok',
          vendor: VENDOR,
          vendor_version: VENDOR_VERSION,
          configured,
          region: s3Config.region,
          endpointUrl: s3Config.endpointUrl || null,
        }));

        if (!configured) return;

        // v1 routes
        await scope.register(async (v1) => {
          // ── List all buckets ──
          v1.get('/buckets', async (_req, reply) => {
            const start = performance.now();
            try {
              const response = await s3Client.send(new ListBucketsCommand({}));
              const buckets = (response.Buckets || []).map((b) => ({
                name: b.Name,
                creationDate: b.CreationDate?.toISOString() || null,
              }));
              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                buckets,
              };
            } catch (err) {
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Create bucket ──
          v1.post('/buckets/:bucket', async (req, reply) => {
            const { bucket } = req.params;
            const start = performance.now();
            try {
              await s3Client.send(new CreateBucketCommand({ Bucket: bucket }));
              reply.code(201);
              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                created: true,
              };
            } catch (err) {
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Check bucket exists ──
          v1.head('/buckets/:bucket', async (req, reply) => {
            const { bucket } = req.params;
            try {
              await s3Client.send(new HeadBucketCommand({ Bucket: bucket }));
              reply.code(200).send();
            } catch {
              reply.code(404).send();
            }
          });

          // ── List objects in bucket ──
          v1.get('/buckets/:bucket/objects', async (req, reply) => {
            const { bucket } = req.params;
            const prefix = req.query.prefix || '';
            const cursor = req.query.cursor || undefined;
            const maxKeys = Math.min(200, Math.max(1, parseInt(req.query.maxKeys, 10) || 50));
            const start = performance.now();

            try {
              const params = { Bucket: bucket, MaxKeys: maxKeys };
              if (prefix) params.Prefix = prefix;
              if (cursor) params.ContinuationToken = cursor;

              const response = await s3Client.send(new ListObjectsV2Command(params));

              const objects = (response.Contents || []).map((obj) => ({
                key: obj.Key,
                size: obj.Size,
                lastModified: obj.LastModified?.toISOString() || null,
                storageClass: obj.StorageClass || null,
              }));

              const commonPrefixes = (response.CommonPrefixes || []).map((p) => p.Prefix);

              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                prefix: prefix || null,
                nextCursor: response.NextContinuationToken || null,
                isTruncated: response.IsTruncated || false,
                keyCount: response.KeyCount || 0,
                objects,
                commonPrefixes,
              };
            } catch (err) {
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Put object (create/update) ──
          v1.post('/buckets/:bucket/objects', async (req, reply) => {
            const { bucket } = req.params;
            const { key, data, contentType } = req.body || {};
            const start = performance.now();

            if (!key) {
              reply.code(400);
              return { status: 'error', error: 'key is required in request body' };
            }

            try {
              const body = typeof data === 'string' ? data : JSON.stringify(data ?? {});
              await s3Client.send(new PutObjectCommand({
                Bucket: bucket,
                Key: key,
                Body: body,
                ContentType: contentType || 'application/json',
              }));

              reply.code(201);
              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                key,
                size_bytes: Buffer.byteLength(body, 'utf-8'),
              };
            } catch (err) {
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Get object by key ──
          v1.get('/buckets/:bucket/objects/by-key', async (req, reply) => {
            const { bucket } = req.params;
            const objectKey = req.query.key;
            const start = performance.now();

            if (!objectKey) {
              reply.code(400);
              return { status: 'error', error: 'query parameter "key" is required' };
            }

            try {
              const response = await s3Client.send(
                new GetObjectCommand({ Bucket: bucket, Key: objectKey }),
              );
              const content = await response.Body.transformToString();

              let parsed;
              try {
                parsed = JSON.parse(content);
              } catch {
                parsed = content;
              }

              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                key: objectKey,
                contentType: response.ContentType || null,
                contentLength: response.ContentLength || 0,
                data: parsed,
              };
            } catch (err) {
              if (isNotFoundError(err)) {
                reply.code(404);
                return { status: 'error', error: `Object not found: ${objectKey}` };
              }
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Update object by key ──
          v1.put('/buckets/:bucket/objects/by-key', async (req, reply) => {
            const { bucket } = req.params;
            const objectKey = req.query.key;
            const start = performance.now();

            if (!objectKey) {
              reply.code(400);
              return { status: 'error', error: 'query parameter "key" is required' };
            }

            try {
              const body = typeof req.body === 'string' ? req.body : JSON.stringify(req.body ?? {});
              await s3Client.send(new PutObjectCommand({
                Bucket: bucket,
                Key: objectKey,
                Body: body,
                ContentType: req.headers['content-type'] || 'application/json',
              }));

              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                key: objectKey,
                size_bytes: Buffer.byteLength(body, 'utf-8'),
              };
            } catch (err) {
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Delete object by key ──
          v1.delete('/buckets/:bucket/objects/by-key', async (req, reply) => {
            const { bucket } = req.params;
            const objectKey = req.query.key;
            const start = performance.now();

            if (!objectKey) {
              reply.code(400);
              return { status: 'error', error: 'query parameter "key" is required' };
            }

            try {
              await s3Client.send(
                new DeleteObjectCommand({ Bucket: bucket, Key: objectKey }),
              );

              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                key: objectKey,
                deleted: true,
              };
            } catch (err) {
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          // ── Get object metadata ──
          v1.get('/buckets/:bucket/objects/metadata', async (req, reply) => {
            const { bucket } = req.params;
            const objectKey = req.query.key;
            const start = performance.now();

            if (!objectKey) {
              reply.code(400);
              return { status: 'error', error: 'query parameter "key" is required' };
            }

            try {
              const response = await s3Client.send(
                new HeadObjectCommand({ Bucket: bucket, Key: objectKey }),
              );

              return {
                status: 'ok',
                latency_ms: Math.round(performance.now() - start),
                bucket,
                key: objectKey,
                contentType: response.ContentType || null,
                contentLength: response.ContentLength || 0,
                lastModified: response.LastModified?.toISOString() || null,
                eTag: response.ETag || null,
                storageClass: response.StorageClass || null,
                metadata: response.Metadata || {},
                serverSideEncryption: response.ServerSideEncryption || null,
              };
            } catch (err) {
              if (isNotFoundError(err)) {
                reply.code(404);
                return { status: 'error', error: `Object not found: ${objectKey}` };
              }
              reply.code(500);
              return { status: 'error', latency_ms: Math.round(performance.now() - start), error: err.message };
            }
          });

          return Promise.resolve();
        }, { prefix: '/v1' });

        return Promise.resolve();
      },
      { prefix: PREFIX },
    );

    server.addHook('onClose', async () => {
      server.log.info('[s3_api] Cleaning up S3 API resources...');
      if (s3Client) {
        try { s3Client.destroy(); } catch { /* ignore */ }
      }
    });

    if (configured) {
      server.log.info({ PREFIX }, `[lifecycle:s3_api] S3 API initialized -- routes registered at ${PREFIX}/*`);
    } else {
      server.log.info({ PREFIX }, `[lifecycle:s3_api] Health endpoint registered at ${PREFIX}/health (v1 routes skipped -- no S3 client)`);
    }

  } catch (err) {
    server.log.error({ err, hookName: '535.s3_api' }, '[lifecycle:s3_api] S3 API initialization failed');
    throw err;
  }
}

/**
 * Shutdown hook.
 * @param {import('fastify').FastifyInstance} server
 */
export async function onShutdown(server) {
  server.log.info('[lifecycle:s3_api] S3 API shutdown complete');
}
